### PR TITLE
Add directory transfer support for SFTPOperator

### PR DESCRIPTION
**Pull request by Dawnpool**
This PR implements directory transfer for SFTPOperator, related to the [issue](https://github.com/apache/airflow/issues/40365) I have raised before.
Currently, the SFTPOperator only accepts file paths, and you have to specify every filename in a folder by list when transferring an entire folder.
By adding some directory handling logic, the operator now can accept directory paths as well, allowing users easily transfer entire folders.

### CODE REVIEW

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (kunaljubce) [inline providers/src/airflow/providers/sftp/hooks/sftp.py]: @Dawnpool Can we add a check if `local_full_path` exists before trying to create it again?

**Comment by kunaljubce** [comment]
@Dawnpool You seem to have a whole test non-db test suite failing. Can you check it on your local breeze and see if you can fix it?

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (Dawnpool) [reply providers/src/airflow/providers/sftp/hooks/sftp.py]: @kunaljubce 
Added it by raising an error if the path already exists. Or, how about just deleting the whole directory and recreating it if it already exists?

**Comment by Dawnpool** [comment]
> @Dawnpool You seem to have a whole test non-db test suite failing. Can you check it on your local breeze and see if you can fix it?

Fixed it by merging the main branch!

**Comment by Dawnpool** [comment]
> @Dawnpool You seem to have a whole test non-db test suite failing. Can you check it on your local breeze and see if you can fix it?

Hi, I guess there was an issue with the test code at the time I forked the repository. It might have been resolved by merging the main branch. There is no problem on my local breeze environment for now.

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (kunaljubce) [inline providers/src/airflow/providers/sftp/hooks/sftp.py:292]: @Dawnpool What happens if this `remote_full_path` is present in the remote location? Did you test this? Maybe add another check here?

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (kunaljubce) [reply providers/src/airflow/providers/sftp/hooks/sftp.py]: @Dawnpool Yeah I am not too much in favour of deleting the existing directory, better left to the user to handle that. Seems fine to raise the error though and stop the process!

**Comment by kunaljubce** [comment]
> > @Dawnpool You seem to have a whole test non-db test suite failing. Can you check it on your local breeze and see if you can fix it?
> 
> Hi, I guess there was an issue with the test code at the time I forked the repository. It might have been resolved by merging the main branch. There is no problem on my local breeze environment for now.

Let's wait for one of the maintainers to approve the remaining workflows. That will give you some clarity if rebasing against `main` fixed it. Feel free to drop a gentle reminder on the Slack channel - `#contributors`.

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (Dawnpool) [reply providers/src/airflow/providers/sftp/hooks/sftp.py:292]: For now, it would copy the elements in the existing remote location. Let me add another error raising here. Thanks for a good catch.

File: providers/src/airflow/providers/sftp/hooks/sftp.py
Comment (Dawnpool) [reply providers/src/airflow/providers/sftp/hooks/sftp.py]: @kunaljubce I've also updated this check from `isdir()` to `exists()` to make sure both a file and a directory with the same name do not exist.

**Comment by potiuk** [comment]
> Let's wait for one of the maintainers to approve the remaining workflows. That will give you some clarity if rebasing against `main` fixed it. Feel free to drop a gentle reminder on the Slack channel - `#contributors`.

Approved workflows.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
errrors, errors everywhere :D

**Comment by kunaljubce** [comment]
@Dawnpool Try reproducing your CI tests locally and figuring out what's going wrong. Seems this is a good place to start - https://github.com/apache/airflow/blob/main/dev/breeze/doc/ci/08_running_ci_locally.md

**Comment by eladkal** [comment]
PR has conflicts that needs to be resolved

**Comment by Dawnpool** [comment]
@potiuk 
The non-db tests keep failing and it doesn't seem related to the changes I made based on the error logs.
I noticed your comment(https://github.com/apache/airflow/pull/44625#issuecomment-2522299932) on a recent approved PR with the same error, where you mentioned that it was an unrelated intermittent error. Is there an ongoing issue with the non-db tests?

@eladkal 
I have resolved the conflicts. Please feel free to check!

**Comment by potiuk** [comment]
I think it's a side effect of bad  configuration - likely some initialization issue in some of the skipped provider tests that are missing somewhere. Likely it is mitigated when full tests are run as some other tests are intitializing SQL Alchemy (where we completely should not need it for DB Tests). I will take a look at it later today and try to find out what it is (And it would be great to not merge it till then as we might be eble to see if the problem is fixed when I find it.

**Comment by potiuk** [comment]
Can you please rebase/resolve conflicts and ping me if it fails again.

**Comment by Dawnpool** [comment]
@potiuk 
it failed again😥 same error with the non-db tests

**Comment by potiuk** [comment]
:scream:

**Comment by potiuk** [comment]
I think this should fix it: https://github.com/apache/airflow/pull/45244

**Comment by potiuk** [comment]
Let's see - I added the fix on top and we should see if the problem is fixed.

**Comment by potiuk** [comment]
Nope - there is another issue

**Comment by potiuk** [comment]
All right. I think I found it https://github.com/apache/airflow/pull/45249 - very interesting issue (and your change accidentally revealed it because it selectively run non-db `microsoft.azure` test collection as the first group of tests to run - and it turned out that those tests relied on a side-effect of other tests being run before.

**Comment by potiuk** [comment]
OK. The fix is merged. You can rebase again and I :crossed_fingers: that it should work now.

**Comment by potiuk** [comment]
Finally!

**Comment by Dawnpool** [comment]
Finally! Thank you all😁

**Comment by Dev-iL** [comment]
Thank you for your contribution @Dawnpool!

In this implementation, files are downloaded sequentially, right? If so - when downloading folders containing a large number of small files, this approach will be quite inefficient. I think we should consider expanding the method, or adding another one that downloads files concurrently.  

Possibly related to [SFTPHookAsync](https://airflow.apache.org/docs/apache-airflow-providers-sftp/stable/_api/airflow/providers/sftp/hooks/sftp/index.html#airflow.providers.sftp.hooks.sftp.SFTPHookAsync).

**Comment by Dawnpool** [comment]
Hi @Dev-iL!
Yes, just as you said, files are downloaded sequentially since it simply uses multiple calls to the `retrieve_file` function in the order they are collected. I totally agree with your point this approach would be inefficient in the case you said.
It would be awesome if we could download multiple files concurrently, but I'm not sure what kind of approach to take (probably multi-threading?) and we should definitely do performance checks to compare after implementing the method.

**Pull request merged**
